js/packages/openinference-instrumentation-langchain/src/utils.ts

```diff
@@ -0,0 +1,299 @@
+import { isAttributeValue } from "@opentelemetry/core";
+import {
+  MimeType,
+  OpenInferenceSpanKind,
+  SemanticConventions,
+} from "@arizeai/openinference-semantic-conventions";
+import { Run } from "@langchain/core/tracers/base";
+import {
+  GenericFunction,
+  LLMMessage,
+  LLMMessageFunctionCall,
+  LLMMessageToolCalls,
+  LLMMessagesAttributes,
+  SafeFunction,
+} from "./types";
+
+export function withSafety<T extends GenericFunction>(fn: T): SafeFunction<T> {
+  return (...args) => {
+    try {
+      return fn(...args);
+    } catch (error) {
+      diag.error(`Failed to get attributes for span: ${error}`);
+      return null;
+    }
+  };
+}
+
+/**
+ * Flattens nested attributes into a single level object
+ */
+function flattenAttributes(
+  attributes: Record<string, unknown>,
+  baseKey: string = "",
+): Attributes {
+  const result: Attributes = {};
+  for (const key in attributes) {
+    const newKey = baseKey ? `${baseKey}.${key}` : key;
+    const value = attributes[key];
+
+    if (value == null) {
+      continue;
+    }
+
+    if (isObject(value)) {
+      Object.assign(result, flattenAttributes(value, newKey));
+    } else if (Array.isArray(value)) {
+      value.forEach((item, index) => {
+        if (isObject(item)) {
+          Object.assign(result, flattenAttributes(item, `${newKey}.${index}`));
+        } else {
+          result[`${newKey}.${index}`] = item;
+        }
+      });
+    } else if (isAttributeValue(value)) {
+      result[newKey] = value;
+    }
+  }
+  return result;
+}
+
+function getOpenInferenceSpanKindFromRunType(runType: string) {
+  const normalizedRunType = runType.toUpperCase();
+  if (normalizedRunType.includes("AGENT")) {
+    return OpenInferenceSpanKind.AGENT;
+  }
+
+  if (normalizedRunType in OpenInferenceSpanKind) {
+    return OpenInferenceSpanKind[
+      normalizedRunType as keyof typeof OpenInferenceSpanKind
+    ];
+  }
+  return "UNKNOWN";
+}
+
+function formatIO({
+  io,
+  type,
+}: {
+  io: Run["inputs"] | Run["outputs"];
+  type: "input" | "output";
+}) {
+  let valueAttribute: string;
+  let mimeTypeAttribute: string;
+  switch (type) {
+    case "input": {
+      valueAttribute = SemanticConventions.INPUT_VALUE;
+      mimeTypeAttribute = SemanticConventions.INPUT_MIME_TYPE;
+      break;
+    }
+    case "output": {
+      valueAttribute = SemanticConventions.OUTPUT_VALUE;
+      mimeTypeAttribute = SemanticConventions.OUTPUT_MIME_TYPE;
+      break;
+    }
+    default:
+      assertUnreachable(type);
+  }
+  if (io == null) {
+    return {};
+  }
+  const values = Object.values(io);
+  if (values.length === 1 && typeof values[0] === "string") {
+    return {
+      [valueAttribute]: values[0],
+      [mimeTypeAttribute]: MimeType.TEXT,
+    };
+  }
+
+  return {
+    [valueAttribute]: JSON.stringify(io),
+    [mimeTypeAttribute]: MimeType.JSON,
+  };
+}
+
+function getRoleFromMessageData(
+  messageData: Record<string, unknown>,
+): string | undefined {
```

Comment:
```suggestion
): ?string {
```

Reply:
any reason we wouldn't provide a fallback?

Reply:
Yeah so my thought was with the undefined and this applies to the comment below as well was that if there is no role or content I don't want it to show up in the attributes as opposed to showing up as null, I just wanted the key to not exist given the logic on line 211 and 217. I can refactor though to do null checks on the role / content and only add the key if not null.

In terms of fallback, I was assuming we only wanted to put things in attributes that actually existed a fallback seems like it wouldn't make sense in most contexts like other models where the roles are inline

Reply:
makes sense. no fallback it is.